File: src/drivecatalog/api/models/folder_duplicate.py
"""Pydantic models for folder-duplicate API responses."""

from __future__ import annotations

from pydantic import BaseModel


class FolderInfo(BaseModel):
    """A folder location with metadata."""

    drive_id: int
    drive_name: str
    folder_path: str
    file_count: int
    total_bytes: int


class ExactMatchGroup(BaseModel):
    """A group of folders with identical file-hash sets."""

    match_type: str  # always "exact"
    hash_count: int
    folders: list[FolderInfo]


class SubsetPair(BaseModel):
    """A pair where one folder's hashes are a proper subset of another's."""

    match_type: str  # always "subset"
    subset_hash_count: int
    superset_hash_count: int
    overlap_percent: float
    subset_folder: FolderInfo
    superset_folder: FolderInfo


class FolderDuplicateStats(BaseModel):
    """Summary statistics for folder duplicate analysis."""

    total_folders_analyzed: int
    exact_match_groups: int
    subset_pairs_found: int


class FolderDuplicateResponse(BaseModel):
    """Top-level response for GET /folder-duplicates."""

    exact_match_groups: list[ExactMatchGroup]
    subset_pairs: list[SubsetPair]
    stats: FolderDuplicateStats

File: DriveCatalog/Models/FolderDuplicate.swift
import Foundation

// MARK: - Folder Duplicate Response

struct FolderDuplicateResponse: Codable {
    let exactMatchGroups: [ExactMatchGroup]
    let subsetPairs: [SubsetPair]
    let stats: FolderDuplicateStats

    enum CodingKeys: String, CodingKey {
        case exactMatchGroups = "exact_match_groups"
        case subsetPairs = "subset_pairs"
        case stats
    }
}

struct FolderInfo: Codable, Identifiable {
    let driveId: Int
    let driveName: String
    let folderPath: String
    let fileCount: Int
    let totalBytes: Int64

    var id: String { "\(driveId):\(folderPath)" }

    enum CodingKeys: String, CodingKey {
        case driveId = "drive_id"
        case driveName = "drive_name"
        case folderPath = "folder_path"
        case fileCount = "file_count"
        case totalBytes = "total_bytes"
    }
}

struct ExactMatchGroup: Codable, Identifiable {
    let matchType: String
    let hashCount: Int
    let folders: [FolderInfo]

    var id: String {
        folders.map(\.id).sorted().joined(separator: "|")
    }

    enum CodingKeys: String, CodingKey {
        case matchType = "match_type"
        case hashCount = "hash_count"
        case folders
    }
}

struct SubsetPair: Codable, Identifiable {
    let matchType: String
    let subsetHashCount: Int
    let supersetHashCount: Int
    let overlapPercent: Double
    let subsetFolder: FolderInfo
    let supersetFolder: FolderInfo

    var id: String { "\(subsetFolder.id)->\(supersetFolder.id)" }

    enum CodingKeys: String, CodingKey {
        case matchType = "match_type"
        case subsetHashCount = "subset_hash_count"
        case supersetHashCount = "superset_hash_count"
        case overlapPercent = "overlap_percent"
        case subsetFolder = "subset_folder"
        case supersetFolder = "superset_folder"
    }
}

struct FolderDuplicateStats: Codable {
    let totalFoldersAnalyzed: Int
    let exactMatchGroups: Int
    let subsetPairsFound: Int

    enum CodingKeys: String, CodingKey {
        case totalFoldersAnalyzed = "total_folders_analyzed"
        case exactMatchGroups = "exact_match_groups"
        case subsetPairsFound = "subset_pairs_found"
    }
}
